compiler/lib/driver.ml

```diff
@@ -105,8 +109,14 @@ let exact_calls profile p =
       | O1 | O2 -> true
     in
     let info = Global_flow.f ~fast p in
-    Specialize.f ~function_arity:(fun f -> Global_flow.function_arity info f) p
-  else p
+    let p =
+      if Config.Flag.globaldeadcode ()
```

Comment:
We don't perform the optimization if `Config.Flag.effects ()` is true, is it expected ?

Reply:
I really only put it here since this is where `Global_flow.f` is called in `driver`, but I'm not sure if it's correct beyond that.

Reply:
global_flow is also called in effects.ml

We should probably move some "driver" logic outside effects.ml into driver.ml

Reply:
We have not looked at that so far since it seemed a bit tricky.
But maybe it just work there as well if the deadcode elimination phase is performed right after the global flow analysis.

Reply:
I've tried to naively apply the global DCE pass when effects are enabled.

Doing DCE before the effect logic result in a bunch of diffs as the following one
```diff
 |       return caml_callback
 |               (function(cont){
 |                 var
+|                  _c_ = undefined,
 |                  global_data = runtime.caml_get_global_data(),
 |                  Stdlib_Printf = global_data.Stdlib__Printf,
 |                  _a_ =
 |                    [0,
 |                     [11, caml_string_of_jsbytes("abc"), 0],
 |                     caml_string_of_jsbytes("abc")];
 |                 function g(param, cont){
-|                  return caml_cps_call2(Stdlib_Printf[2], _a_, cont);
+|                  var _g_ = Stdlib_Printf[2];
+|                  return caml_cps_call2
+|                          (_g_, _a_, function(_h_){return cont(_c_);});
 |                 }
-|                 caml_callback(g, [0]);
+|                 caml_callback(g, [_c_]);
```

Doing DCE after the effect logic gives me better diff but also triggers compilation and runtime errors.
```diff
------ compiler/tests-compiler/lambda_lifting.ml
++++++ compiler/tests-compiler/lambda_lifting.ml.corrected
File "compiler/tests-compiler/lambda_lifting.ml", line 18, characters 0-1:
 |  let prog =
 |    {|
 |let f x =
 |  let g y =
 |    let h z = x + y + z in
 |    h 7
 |  in
 |  g 5
 |in
 |Printf.printf "%d\n" (f 3)
 |    |}
 |  in
 |  let flags =
 |    [ "--no-inline"; "--set=lifting-threshold=1"; "--set=lifting-baseline=0" ]
 |  in
 |  Util.compile_and_run ~effects:true ~flags prog;
-|  [%expect {|15 |}];
+|  [%expect {|
+|    /tmp/build_259527_dune/jsoo-test4651ac/test.js:2741
+|         throw err;
+|         ^
+|
+|    TypeError: Cannot read properties of undefined (reading 'opened')
+|        at caml_ml_output_ta (/tmp/build_259527_dune/jsoo-test4651ac/test.js:1837:15)
+|        at caml_ml_output_bytes (/tmp/build_259527_dune/jsoo-test4651ac/test.js:1876:12)
+|        at caml_ml_output (/tmp/build_259527_dune/jsoo-test4651ac/test.js:1879:12)
+|        at output_string (/tmp/build_259527_dune/jsoo-test4651ac/test.js:3146:5)
+|        at /tmp/build_259527_dune/jsoo-test4651ac/test.js:5523:27
+|        at _cj_ (/tmp/build_259527_dune/jsoo-test4651ac/test.js:5540:41)
+|        at output_acc (/tmp/build_259527_dune/jsoo-test4651ac/test.js:5511:58)
+|        at caml_cps_exact_call3 (/tmp/build_259527_dune/jsoo-test4651ac/test.js:2792:15)
+|        at _cj_ (/tmp/build_259527_dune/jsoo-test4651ac/test.js:5568:16)
+|        at output_acc (/tmp/build_259527_dune/jsoo-test4651ac/test.js:5511:58)
+|
+|    Node.js v20.0.0
+|
+|    process exited with error code 7
+|     node test.js |}];
```

and 

```
  _build/default/compiler/bin-js_of_ocaml/js_of_ocaml.exe: Error: Invalid_argument("List.iter2")
```

Reply:
The global DCE does not preserve tail calls. We don't care normally since JavaScript does not support tail calls, but this makes a difference for the CPS transformation.

Maybe we can disable this transformation when effects are enabled?
https://github.com/ocsigen/js_of_ocaml/blob/8d1841f748ab5ef4a4eb37d4b1e13dddc934b132/compiler/lib/global_deadcode.ml#L355

Reply:
Right now we only call `Global_flow` when effects are disabled, so I'm a little confused where the proposed change for calling global deadcode would be in the case that effects are enabled.

Reply:
Global_flow is called inside effect.ml

Reply:
Thanks, missed that one

Reply:
Okay I also tried a naive effort to call Global_flow with effects [here](https://github.com/micahcantor/js_of_ocaml/pull/1/files) with @vouillon 's suggestion, but, it's still crashing a bunch of effects tests. I don't really know anything about that pass so it's hard to debug.

Reply:
It's hard for me to gauge how large of a change is required for this, from tweaking a few lines to a bigger refactor of the driver. If it's more like the latter, maybe we should split this into a separate PR?

Reply:
I ask only because I have an article I'm waiting to post until this is merged, but haven't had time to take a closer look at this issue. I also don't know of any examples they would benefit from effects + gdc (but there might be!)

Reply:
@vouillon @hhugo

Reply:
If @hhugo, I think we can merge this PR, indeed.

Reply:
It seems it's missing a changelog entry, though.

Reply:
There are still a few points I'd like to look at before merging.
I ll try to do it today

Reply:
Thanks. I'll add a PR for this effects change too